File: lib/dnls/pytorch/nn/temporal_inds.py
"""

Get temporal inds from spatial inds

"""

import torch as th
import dnls_cuda

def run(inds,wt,fflow,bflow):

    # -- shaping --
    if inds.ndim == 4: inds = inds.unsqueeze(1)
    inds = inds.contiguous()
    B,HD,Q,K,_ = inds.shape
    B,T,_,H,W = fflow.shape
    st = 2*wt
    nT = min(st,T-1)

    # -- allocate --
    inds_t = th.ones((B,HD,Q,K,nT,3),device=inds.device,dtype=inds.dtype)
    inds_t[...] = -1

    # -- run --
    dnls_cuda.temporal_inds(inds,fflow,bflow,inds_t)
    th.cuda.synchronize()

    return inds_t

File: lib/csrc/dev/nn/temporal_inds.cpp
// -- included for pytorch --
#include <torch/extension.h>
#include <vector>

// -- include cuda_runtime for jax --
#include <cuda_runtime_api.h>
#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <pybind11/pybind11.h>
#include "../jax_pybind.h"
using namespace torch::indexing;


// CUDA forward declarations

void temporal_inds_cuda(
    const torch::Tensor inds,
    const torch::Tensor fflow,
    const torch::Tensor bflow,
    torch::Tensor inds_t);

// C++ interface

#define CHECK_CUDA(x) TORCH_CHECK(x.device().is_cuda(), #x " must be a CUDA tensor")
#define CHECK_CONTIGUOUS(x) TORCH_CHECK(x.is_contiguous(), #x " must be contiguous")
#define CHECK_INPUT(x) CHECK_CUDA(x); CHECK_CONTIGUOUS(x)

void temporal_inds(
    const torch::Tensor inds,
    const torch::Tensor fflow,
    const torch::Tensor bflow,
    torch::Tensor inds_t){
  CHECK_INPUT(inds);
  CHECK_INPUT(fflow);
  CHECK_INPUT(bflow);
  CHECK_INPUT(inds_t);
  temporal_inds_cuda(inds, fflow, bflow, inds_t);
}



// python bindings
void init_temporal_inds(py::module &m){
  m.def("temporal_inds", &temporal_inds,
        "Get the Temporal Index Offsets using Opitcal Flow (CUDA)");
}
